File: upvpn-android/app/src/main/kotlin/app/upvpn/upvpn/data/VPNSessionWatcher.kt
package app.upvpn.upvpn.data

import android.util.Log
import app.upvpn.upvpn.data.db.VPNDatabase
import app.upvpn.upvpn.model.VpnSessionStatus
import app.upvpn.upvpn.model.VpnSessionStatusRequest
import com.github.michaelbull.result.onFailure
import com.github.michaelbull.result.onSuccess
import kotlinx.coroutines.delay

class VPNSessionWatcher(
    private val request: VpnSessionStatusRequest,
    private val vpnDatabase: VPNDatabase,
    private val vpnSessionRepository: VPNSessionRepository
) {
    private val tag = "VPNSessionWatcher"
    suspend fun watch(block: (VpnSessionStatus) -> Unit) {
        var done = false
        while (done.not()) {
            Log.i(tag, "watching ...")
            delay(1000)
            val status = vpnSessionRepository.getVpnSessionStatus(request)

            status.onSuccess(block)
            // if the status is end state break
            status.onSuccess {
                if (it is VpnSessionStatus.ServerReady || it is VpnSessionStatus.Failed || it is VpnSessionStatus.Ended) {
                    done = true
                }
            }

            status.onSuccess {
                if (it is VpnSessionStatus.Failed || it is VpnSessionStatus.Ended) {
                    vpnDatabase.vpnSessionDao().deleteWithRequestId(request.requestId)
                }
            }

            status.onFailure { Log.i(tag, "Failed to get status in watcher $it") }

        }
        Log.i(tag, "watch ended")
    }
}

File: upvpn-apple/UpVPN/UpVPNNetworkExtension/Data/VPNSessionWatcher.swift
//
//  VPNSessionWatcher.swift
//  UpVPN
//
//  Created by Himanshu on 7/10/24.
//

import Foundation
import os.log

class VPNSessionWatcher {
    private var request: VpnSessionStatusRequest
    private var vpnSessionRepository: VPNSessionRepository
    private var location: Location
    private var done = false
    private var onStatusUpdate: (VpnSessionStatus, Location) -> Void

    init(request: VpnSessionStatusRequest,
         vpnSessionRepository: VPNSessionRepository,
         location: Location,
         onStatusUpdate: @escaping (VpnSessionStatus, Location) -> Void) {
        self.request = request
        self.vpnSessionRepository = vpnSessionRepository
        self.location = location
        self.onStatusUpdate = onStatusUpdate
    }

    func watch() -> Task<Void, Never> {
        let task = Task.detached { [self] in
            os_log("watcher started")
            while !self.done && !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 1_000_000_000)
                
                let result = await vpnSessionRepository.getVpnSessionStatus(request: self.request)

                guard case .success(let status) = result else {
                    // todo log error
                    continue
                }

                self.onStatusUpdate(status, location)

                // if status is a terminal state end watcher
                switch status {
                case .failed, .serverReady, .clientConnected, .ended:
                    self.done = true
                default:
                    continue
                }
            }
            os_log("watcher stopped")
        }
        return task
    }
}
